### ISSUE TITLE
Rule proposal: Forbid reassign top level variable in functions

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1092]
It's dangerous, and maybe get unexpected result when refactoring. In following example, if we change `foo` to `async`, or inside `foo` somehow reenter `foo`, `cache` may got change.

## Fail

```js
let cache; 

function foo() {
	cache = {};
	// ...
}

export default foo;
```

## Pass

```js
function foo() {
	let cache;

	// ...
}
```

**Comment by fisker** [comment]
You may ask why the failed example write in the way as passed example, here is the story

```js
function bar(x) {
	if (x.array) {
		x.array = x.array.map(bar);
	}

	return x;
}

function foo(x) {
	// ...
	return bar(x);
}
```

I want add `id` to `x`

```js
let cache;

function getId(y) {
	if (cache[y]) {
		return y;
	}

	return Symbol();
}

function bar(x) {
	x.id = getId(y); // newly added
	if (x.array) {
		x.array = x.array.map(bar);
	}

	return x;
}

function foo(x) {
	cache = {};

	// ...
	return bar(x);
}
```

This is a easy way.

Problem in this case is, if `foo` use a local variable `cache`,  `getId` function have to move inside `foo`, then `bar` function have to... in real world case, there might be more functions need move into `foo`.

**Comment by sindresorhus** [comment]
I have sometimes needed to reassign a top-level variable for caching purposes, but it's so seldom, I'm ok with just using an `eslint-disable` comment in such case.

**Comment by sindresorhus** [comment]
Any thoughts on the rule name? `no-top-level-assignment-in-function`?

**Comment by fisker** [comment]
I can't think a better name.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by futpib** [comment]
Should this be limited to the top level variables or any upper level variables?
